### PR TITLE
Build api for transaction order

### PR DESCRIPTION
**Pull request by ledacluc**
just built Created, Please allow me to add it tomorrow evening.

### CODE REVIEW

File: server/controllers/authController.js [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+                    success: false,
+                    result: null,
+                    message: 'Password need to be at least 8 characters long',
+                })
+
+            //MA HOA PASSWORD 
+            const hashedPassword = await bcrypt.hash(password, 10)
+            req.body.password = hashedPassword;
+            //Create new user
+            const result = await new User(req.body).save({});
```
Comment (affkjuhyt) [inline server/controllers/authController.js:56]:
```
const user = new User(req.body)
const result = await user.save()
```

File: server/controllers/authController.js [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+                });
+            }
+            return res.status(200).send({
+                success: true,
+                result: {
+                    _id: result._id,
+                    name: result.name,
+                    email: result.email,
+                    phone: result.phone,
+                    password: hashedPassword,
```
Comment (affkjuhyt) [inline server/controllers/authController.js:71]: Not show hashed password

File: server/controllers/authController.js [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+                return res.status(200).json({
+                    success: true,
+                    result: {
+                        _id: user._id,
+                        name: user.name,
+                        email: user.email,
+                        phone: user.phone,
+                        role: user.role,
+                        address: user.address,
+                        Access_Token, Refresh_Token
```
Comment (affkjuhyt) [inline server/controllers/authController.js:138]:
```
token: {
   access_token: value,
   refress_token: value
}
```

File: server/controllers/orderController.js [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        try {
+            const { orderItems, paymentMethod, itemsPrice, shippingPrice, totalPrice, fullName, user, address, phone } = req.body;
+            if (!orderItems || !paymentMethod || !itemsPrice || !shippingPrice || !totalPrice || !fullName || !address, !phone) {
+                return res.status(400).json({
+                    result: null,
+                    success: false,
+                    message: 'the input is required.',
+                })
+            }
+            const promises = orderItems.map(async (order) => {
```
Comment (affkjuhyt) [inline server/controllers/orderController.js:18]: - Insert data to Order table
- Update field quantity of Product
- Transaction

File: server/controllers/userController.js [hunk 5]
```diff
@@ -1,29 +1,45 @@
+const bcrypt = require("bcryptjs");
+// const jwt = require('jsonwebtoken');
 const User = require('../models/userModel');
 
 //ADD User
-exports.createUser = async (req, res) => {
+exports.registerUser = async (req, res, next) => {
```
Comment (affkjuhyt) [inline server/controllers/userController.js:6]: createUser

File: server/controllers/userController.js [hunk 6]
```diff
@@ -51,28 +67,50 @@ exports.createUser = async (req, res) => {
...
     res.status(501).json({ success: false, message: 'error server' });
   }
 },
 
-  //GET all User
-  exports.getUsers = async (req, res) => {
-    console.log('get users');
+
+  //LOGIN USER
+  exports.loginUser = async (req, res) => {
```
Comment (affkjuhyt) [inline server/controllers/userController.js:85]: login

File: server/index.js [hunk 7]
```diff
@@ -33,7 +35,8 @@ app.get('/', (req, res) => {
 });
 app.use("/api/v1/users", userRouter);
 app.use("/api/v1/products", productRouter);
-
+app.use("/api/v1/oder", orderRouter);
```
Comment (affkjuhyt) [inline server/index.js:38]: Order

File: server/middleware/catchAsyncErrors.js [hunk 8]
```diff
@@ -0,0 +1,3 @@
+module.exports = func => (req, res, next) =>
+    Promise.resolve(func(req, res, next))
+        .catch(next)
```
Comment (affkjuhyt) [inline server/middleware/catchAsyncErrors.js:3]: EOL

File: server/models/productModel.js [hunk 9]
```diff
@@ -5,14 +5,12 @@ const productSchema = new mongoose.Schema(
     {
         name: { type: String, required: true, unique: true },
         price: { type: Number, required: true, default: 0.000 },
-        type: { type: String, required: true },
+        category: { type: String, required: true },
         countInstock: { type: Number, required: true },
```
Comment (affkjuhyt) [inline server/models/productModel.js:9]: count_instock

File: server/models/userModel.js [hunk 10]
```diff
@@ -30,6 +30,7 @@ const userSchema = new Schema(
         address: { type: String, maxlength: 100 },
         role: { type: String, enum: ['user', 'admin'], default: 'user' },
         password: { type: String, required: true },
+        confirmPassword: { type: String, required: true },
```
Comment (affkjuhyt) [inline server/models/userModel.js:33]: confirm_password

File: server/routes/coreRoutes/authRoutes.js [hunk 11]
```diff
@@ -1,108 +1,12 @@
...
-//Middleware validation cho API đăng ký
-const validateSignup = (req, res) => {
-    const { name, email, phone, address, password, confirmPassword } = req.body
+router
 
-    //Kiểm tra các trường có giá trị không
-    if (!name || !email || !phone || !address || !password || !confirmPassword) {
-        return res.status(400).json({ message: 'Vui lòng điền đầy đủ thông tin' })
-    }
+    .post("/register", authController.registerUser) //SignUp User
```
Comment (affkjuhyt) [inline server/routes/coreRoutes/authRoutes.js:9]: signup -> signup

File: server/routes/coreRoutes/authRoutes.js [hunk 12]
```diff
@@ -1,108 +1,12 @@
...
+    .post("/log-in", authController.loginUser) // Log-In User
```
Comment (affkjuhyt) [inline server/routes/coreRoutes/authRoutes.js:10]: signin or login

File: server/routes/coreRoutes/transactionRoutes.js [hunk 13]
```diff
@@ -0,0 +1,8 @@
+const express = require('express');
+const transactionController = require('../../controllers/transactionController');
+
+const router = express.Router()
+
+router.post("/create", transactionController.processPayment);
```
Comment (affkjuhyt) [inline server/routes/coreRoutes/transactionRoutes.js:6]: create -> payment

File: server/routes/coreRoutes/userRoutes.js [hunk 14]
```diff
@@ -4,7 +4,8 @@ const userController = require('../../controllers/userController');
 const router = express.Router()
 
 router
-  .post("/create/", userController.createUser) //Add User
+  .post("/sign-up", userController.createUser) //SignUp User
```
Comment (affkjuhyt) [inline server/routes/coreRoutes/userRoutes.js:7]: signup
login

File: server/models/productModel.js (hunk 9 above)
Comment (ledacluc) [reply server/models/productModel.js:9]: changed

File: server/models/userModel.js (hunk 10 above)
Comment (ledacluc) [reply server/models/userModel.js:33]: changed

File: server/routes/coreRoutes/authRoutes.js (hunk 11 above)
Comment (ledacluc) [reply server/routes/coreRoutes/authRoutes.js:9]: changed

File: server/routes/coreRoutes/authRoutes.js (hunk 12 above)
Comment (ledacluc) [reply server/routes/coreRoutes/authRoutes.js:10]: changed

File: server/routes/coreRoutes/transactionRoutes.js (hunk 13 above)
Comment (ledacluc) [reply server/routes/coreRoutes/transactionRoutes.js:6]: changed

File: server/routes/coreRoutes/userRoutes.js (hunk 14 above)
Comment (ledacluc) [reply server/routes/coreRoutes/userRoutes.js:7]: changed

File: server/index.js (hunk 7 above)
Comment (ledacluc) [reply server/index.js:38]: changed

File: server/controllers/userController.js (hunk 6 above)
Comment (ledacluc) [reply server/controllers/userController.js:85]: changed

File: server/controllers/authController.js (hunk 1 above)
Comment (ledacluc) [reply server/controllers/authController.js:56]: changed

File: server/controllers/authController.js (hunk 2 above)
Comment (ledacluc) [reply server/controllers/authController.js:71]: changed

File: server/controllers/authController.js (hunk 3 above)
Comment (ledacluc) [reply server/controllers/authController.js:138]: changed

**Comment by affkjuhyt** [comment]
Fix conflict rồi đẩy lại lên nhé @ledacluc

File: server/controllers/authController.js [hunk 15]
```diff
@@ -0,0 +1,149 @@
...
+            if (!name || !email || !phone || !password || !confirmPassword)
+                return res.status(400).json({
+                    success: false,
+                    result: null,
+                    message: "Name, Email , Phone or password  fields they don't have been entered.",
+                })
+
+            // Kiểm tra định dạng email   
+            if (!email.match(/^[^\s@]+@[^\s@]+\.[^\s@]+$/)) {
+                return res.status(400).json({ message: 'invalid  email' })
```
Comment (affkjuhyt) [inline server/controllers/authController.js:20]: Đặt lại tên message cho chuẩn ví  dụ "Invalid email" viết hoa chữ cái đầu lên

File: server/routes/coreRoutes/orderRoutes.js [hunk 16]
```diff
@@ -0,0 +1,8 @@
+const express = require('express');
+const orderController = require('../../controllers/orderController');
+
+const router = express.Router()
+
+// router.post("/create", orderController.createProduct);
+
+module.exports = router;
```
Comment (affkjuhyt) [inline server/routes/coreRoutes/orderRoutes.js:8]: EOL

**Review by affkjuhyt — APPROVED** [review]

File: server/routes/coreRoutes/orderRoutes.js (hunk 16 above)
Comment (ledacluc) [reply server/routes/coreRoutes/orderRoutes.js:8]: đã sửa

File: server/middleware/catchAsyncErrors.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (ledacluc) [reply server/middleware/catchAsyncErrors.js:3]: File was deleted because it was not in use

**Pull request merged**